Add yarn-berry3 and yarn-berry4 to fetch and use version 2 yarn.lock files

gador opened the pull request:
### Problem

Currently there is no good way to use newer `yarn.lock` files because they rely on `yarn-berry` and cannot be used together with pre-existing tooling like `fetch-yarn-deps`. 
There is an (abandoned? ) [project](https://github.com/VHT/yarn-lock-converter) which converts `v1` files to `v2` files with some caveats. This does require an internet connection and cannot be used in an isolated builder. 

So one needs to convert the `yarn.lock` file and commit it to `nixpkgs` to be used for an offline cache. 

There is an issue open (https://github.com/NixOS/nixpkgs/issues/254369)  which discusses some ways around it. The most sophisticated way was by @szlend and this PR builds on top of it to extend to be included in our new `yarnConfigHook` from #318015

### Implementation of a solution

A new argument has been added to fetch-yarn-deps: `yarnVersion` which will default to `1` which uses the original `yarn`.

For newer `yarn.lock` files, one can set the `yarnVersion` to either `3` or `4` which will use yarn-berry3 or yarn-berry4 respectively. We need to separate both versions, because they have some small changes in the way the lock file is created.

This also adds the corresponding tests.

### How to use it

For an implementation see the included `pgadmin4` derivation, which was also my main motivation for this. I had to build a rather ugly and complicated update script before and do a lot of custom work to build the frontend. This is now simplified. 

In short, add `yarnBerry3ConfigHook` to `nativeBuildInputs` and 
```
yarnOfflineCache = fetchYarnDeps {
    src = "...";
    yarnVersion = 3;
    hash = "..."
  };
```
to your derivation.

fixes https://github.com/NixOS/nixpkgs/issues/254369

Once we agree on the implantation, I'll add the documentation.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
- [ ] Add documentation

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

gador committed:
yarn-berry_3: init at 3.8.6

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
fetch-yarn-deps: add yarn-berry support

A new argument has been added to fetch-yarn-deps:
`yarnVersion` which will default to `1` which uses
the original `yarn`.

For newer `yarn-berry` `yarn.lock`, one can set the
`yarnVersion` to either `3` or `4` which will use
yarn-berry3 or yarn-berry4 respectively.

The difference is the `yarn.lock` file, which follows
a different format, depending on the version.

This also adds the corresponding tests.

The added support for yarn.lock files > version 1 has been
inspired by @szlend from
https://github.com/NixOS/nixpkgs/issues/254369#issuecomment-2080460150

fixes #254369

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>
Co-authored-by: Doron Behar <doron.behar@gmail.com>

gador committed:
pgadmin4: use new yarnBerry3ConfigHook

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

doronbehar commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -131,23 +134,68 @@ in
-              prefetch-yarn-deps
-              cacert
-            ];
+            nativeBuildInputs =
+              [ cacert ]
+              ++ (
+                if (yarnVersion == 1) then
+                  [ prefetch-yarn-deps ]
+                else if (yarnVersion == 3) then
+                  [ yarn-berry3 ]
+                else if (yarnVersion == 4) then
+                  [ yarn-berry4 ]
+                else
+                  [ ]
+              );
```
Readability improvement:

```suggestion
            nativeBuildInputs = [
              ({
                1 = prefetch-yarn-deps;
                3 = yarn-berry3;
                4 = yarn-berry4;
              }.${yarnVersion})
              cacert
            ];
```

Also, it will naturally error when an unsupported version will be given.

gador replied:
thanks!

doronbehar commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -131,23 +134,68 @@ in
 
-            buildPhase = ''
-              runHook preBuild
+            supportedArchitectures = builtins.toJSON {
+              os = [
+                "darwin"
+                "linux"
+              ];
+              cpu = [
+                "arm"
+                "arm64"
+                "ia32"
+                "x64"
+              ];
+              libc = [
+                "glibc"
+                "musl"
+              ];
+            };
```
It'd be slightly better if changing this variable would not trigger a change of also the `yarnVersion == 1` derivations. It can be achieved by defining this attribute in the passthru, and using `lib.escapeShellArg`.

gador replied:
I don't believe this will change at all.. @szlend this is based on your comment. Do you beliefe this has the potential to change in the foreseeable future?

szlend replied:
I'm not in the position to merge PRs, but personally I wouldn't touch any of the old builders and just start out fresh, taking pnpm as inspiration. Yarn Berry is a completely different package manager with a different API so imo it doesn't make sense to mix it with the old one.

gador replied:
I like that it is behind a unified `fetchYarnDeps`. I first implemented a `fetchYarnDepsBerry` separate function but it 1. looks ugly, 2. doesn't follow the [POLS](https://de.wikipedia.org/wiki/Principle_of_Least_Surprise) 
Although `yarn` and `yarn-berry` are different, they are (IMHO) nicely separated by `yarnVersion`

The API (so, basically the CLI) is nicely hidden behind `fetchYarnDeps` and the new `yarnBerry3ConfigHook`. 

The question here is just, whether the `supportedArchitectures` will change in the future? 
Also, when I ran `nixpkgs-review` on this PR, it did not trigger any mass rebuilds, but only rebuilt a handful of packages. So I believe, even if `supportedArchitectures` would change, it will not trigger a mass rebuild for `v1` packages

szlend replied:
This a fixed-output derivation so the only way to make it rebuild is to change the hash, or forcefully with --rebuild. If supportedArchitectures changed, nothing would actually happen because the deelrivation with this output has was already in the nix store But it would silently make the derivation non-reproducible.

If this value ever needs to change then we need scan nixpkgs and update the output hashes.

We could predefine all possible supported platforms, or just the one nixpkgs currently supports. Not sure what the best option is here.

Sorry I'm afk this week, writing this on my phone so I can't really have a look at this in any depth until I'm back.

gador replied:
This probably won't be an issue:

1. The tests for `fetchYarndeps` still passed. So, although I added the `yarnVersion` variable, I did not have to change their hashes
2. I changed the hash of an unrelated `fetchYarnDeps` from an unrelated package to `lib.fakeHash`, confirmed its the same as before after redownloading all packages, and changed the `supportedArchitectures`. The hash after that was still the same as it was initially. 

So changing`supportedArchitectures` in the future will have no effect on `yarnVersion = 1` packages

doronbehar replied:
Yea you are right... I forgot this is a FOD. Marking conversation as resolved.

doronbehar commented on pkgs/development/tools/yarn-berry/default.nix:
```diff
@@ -18,13 +26,9 @@ stdenv.mkDerivation (finalAttrs: {
-  buildInputs = [
-    nodejs
-  ];
+  buildInputs = [ nodejs ];
```
This diff shouldn't be necessary with the latest nixfmt. See https://github.com/NixOS/nixfmt/pull/224 .

gador replied:
see https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1837038425

doronbehar commented on pkgs/development/tools/yarn-berry/default.nix:
```diff
@@ -43,16 +47,18 @@ stdenv.mkDerivation (finalAttrs: {
-    version = testers.testVersion {
-      package = finalAttrs.finalPackage;
-    };
+    version = testers.testVersion { package = finalAttrs.finalPackage; };
```
I think this change shouldn't be necessary too with https://github.com/NixOS/nixfmt/pull/224 .

gador replied:
currently, the CI does not complain. Should we wait for an update there?

doronbehar replied:
> currently, the CI does not complain. Should we wait for an update there?

It won't complain either way, I'm just an advocate of minimal diffs, and an advocate of always putting `*Inputs` and other attributes, that very often grow/change in future updates of a package.

gador replied:
I tried to apply your suggestions, but `nixfmt` always redid what is seen here. I prefer not to interfere with autoformating.

doronbehar replied:
> I tried to apply your suggestions, but `nixfmt` always redid what is seen here. I prefer not to interfere with autoformating.

Are you sure you are using the latest `nixfmt`? Note also the difference between `nixfmt-classic` and `nixfmt-rfc-style`.

doronbehar replied:
> yes.
> 
> ```
> nixfmt nixpkgs-unstable-2024-08-16
> ```
> 
> which is on current `nixos-unstable` and `master` (and yes thats `nixfmt-rfc-style`)

Not my experience:

https://asciinema.org/a/ynlyFqkX7cjaSOMBquT677rYi

doronbehar commented on pkgs/development/tools/yarn-berry/default.nix:
```diff
@@ -1,4 +1,12 @@
-{ fetchFromGitHub, lib, nodejs, stdenv, testers, yarn, berryVersion ? 4 }:
+{
+  fetchFromGitHub,
+  lib,
+  nodejs,
+  stdenv,
+  testers,
+  yarn,
+  berryVersion ? 4,
+}:
```
It'd be a bit nicer if the formatting change would come before the `yarn-berry` 3 commit. Also, the commit title should be: `yarn-berry3: init at 3.8.6`.

gador replied:
done

doronbehar commented:
Note also the CI failure.

gador commented:
> Note also the CI failure.

This is caused by not adding the `yarn-berry3` and `yarn-berry4` to pkgs/by-name which doesn't make sense here

doronbehar commented:
> > Note also the CI failure.
> 
> This is caused by not adding the `yarn-berry3` and `yarn-berry4` to pkgs/by-name which doesn't make sense here

Hmm seems like a false negative to me. Let's ask the advice of @NixOS/nixpkgs-vet .

willbush commented:
Think this recommendation applies here: https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#recommendation-for-new-packages-with-multiple-versions

gador commented:
> Think this recommendation applies here: https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#recommendation-for-new-packages-with-multiple-versions

Awesome, thanks for the pointer!

doronbehar reviewed (commented):
Looks pretty good, though I wouldn't feel comfortable to merge this because I don't know the `yarn config set`tings in `yarnBerryConfigHook` but they looks pretty reasonable.

gador commented on pkgs/tools/admin/pgadmin/default.nix:
```diff
@@ -46,6 +46,22 @@ let
+    name = "pgadmin4-frontend";
+    src = "${src}/web";
+
+    nativeBuildInputs = [ yarnBerry3ConfigHook ];
```
well, I just used the autoformat..I would leave the formatting as is and let the `nixfmt` do its thing..

gador commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5942,7 +5944,17 @@ with pkgs;
   yapf = with python3Packages; toPythonApplication yapf;
 
-  yarn-berry = callPackage ../development/tools/yarn-berry { };
+  inherit
+    ({
+      yarn-berry3 = callPackage ../development/tools/yarn-berry { berryVersion = 3; };
+
+      yarn-berry4 = callPackage ../development/tools/yarn-berry { };
+    })
+    yarn-berry3
+    yarn-berry4
+    ;
```
Well, theoretically it could work I suppose. Question is, what is the preferred way? According to https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#recommendation-for-new-packages-with-multiple-versions it seems to be preferred with `inherit`  but I'm fine either way

SuperSandro2000 replied:
I would prefer what doronbehar wrote, too.

doronbehar replied:
> Well, theoretically it could work I suppose. Question is, what is the preferred way? According to [`master`/pkgs/by-name#recommendation-for-new-packages-with-multiple-versions](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name?rgh-link-date=2024-11-12T11%3A33%3A40Z#recommendation-for-new-packages-with-multiple-versions) it seems to be preferred with `inherit` but I'm fine either way

Hmm yes but in this case, there are also these instructions: https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#manual-migration-guidelines which I think are a bit more relevant.

gador commented on pkgs/tools/admin/pgadmin/update.sh:
```diff
@@ -52,56 +47,15 @@ wget -c $url
+export YARN_ENABLE_TELEMETRY=0
+yarn config set enableGlobalCache false
+yarn config set cacheFolder cache
+yarn config set supportedArchitectures --json '{ "os": [ "darwin", "linux" ], "cpu": [ "arm", "arm64", "ia32", "x64" ], "libc": [ "glibc", "musl" ] }'
```
Well this is the (bash) update script. If you have an idea how to import a nix formatted list here we can change it?

doronbehar replied:
> Well this is the (bash) update script. If you have an idea how to import a nix formatted list here we can change it?

We can save a json file, and use `"$(cat path2.json)"` here, and do something similar in the hook.

gador replied:
done

gador committed:
yarn-berry: add gador as maintainer

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5942,7 +5944,17 @@ with pkgs;
-  yarn-berry = callPackage ../development/tools/yarn-berry { };
+  inherit
+    ({
+      yarn-berry3 = callPackage ../development/tools/yarn-berry { berryVersion = 3; };
+
+      yarn-berry4 = callPackage ../development/tools/yarn-berry { };
```
done

gador committed:
yarn-berry[_3|_4]: move to pkgs/by-name

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
doc/javascript: document yarn-berry for fetchYarnDeps

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

winterqt commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -132,22 +135,49 @@ in
+                mkdir -p $out
+              ''
+              + lib.optionalString (yarnVersion > 1) ''
+                yarn install --immutable --mode skip-build
```
This will break the FODs almost instantly if Yarn changes anything in how they pull deps, and also is probably not reproducible between platforms when platform-specific dependencies are used.

gador replied:
Valid point. This is partly why there are two yarn versions.

Also this could be said for all external package managers, not just yarn-berry.

I'm uncertain about your point about platform specific dependencies. I'll have to look into it.

szlend replied:
Platform specific dependencies should be made reproducible by explicitly setting `supportedArchitectures`. Your point about Yarn breaking how they fetch deps may be valid depending on the upstream stability guarantees. Explicitly picking the version of yarn somewhat mitigates that issue, but it's no guarantee yes. This is the same path taken in `pnpm.fetchDeps` #290715

winterqt reviewed (changes requested):
This is a no-go in its current state.

doronbehar commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -148,23 +148,6 @@ in
-                "glibc"
-                "musl"
-              ];
-            };
```
Could you squash this `supportedArchitectures` json file commit to the one that added this?

gador replied:
done

doronbehar reviewed (commented):
Small request, looks pretty good overall.

doronbehar commented on doc/languages-frameworks/javascript.section.md:
````diff
@@ -594,6 +594,24 @@ To install the package `yarnInstallHook` uses both `npm` and `yarn` to cleanup p
+The above works well with `yarn` version 1. However, `yarn-berry` introduced a new version of lockfiles, which will not work with classic `yarn`. To still be able to use `fetchYarnDeps` with `yarn-berry`, you will need to supply the source folder containing the `yarn.lock`  and the corresponding project files, and the version of the lockfile:
+
+```nix
+yarnOfflineCache = fetchYarnDeps {
+    src = "${src}/web";
+    yarnVersion = 3;
+    hash = "...";
+  };
````
First of all:

```suggestion
yarnOfflineCache = fetchYarnDeps {
  inherit (finalAttrs) src;
  yarnVersion = 3;
  hash = "...";
};
```

Not only that, I noticed that with `yarnVersion`, there should be a `yarnLock` argument to `fetchYarnDeps`, and not `src` argument like here. Do you think you can modify `fetchYarnDeps` a bit such that both usages of the function will get a `yarnLock` argument?

gador replied:
It actually needs more than the `lock` file (the `package.json` is mandatory, too, see the test folders)

gador replied:
mmh, maybe thats not a bad idea. So we'll use the `yarnLock` argument and derive the directory name from it and use this (internally) as a `src`

doronbehar replied:
> mmh, maybe thats not a bad idea. So we'll use the `yarnLock` argument and derive the directory name from it and use this (internally) as a `src`

Exactly that's what I suggest.

gador replied:
Ok, I gave it a try, but this is actually harder than I thought. The reason is, that the `yarnLock` file is a _singular_  file in the nixstore, completely independent of the project folder. `yarn-berry` needs both files (it actually wants the project folder, but these two files suffice), but we cannot deduce the project folder from a yarn lockfile which is just a store path. We _need_ a directory (or a separate `packageJson` argument)

gador replied:
Small update. I would work with 
```nix
berry_4 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
    yarnLock = "${./berry_4}/yarn.lock";
    yarnVersion = 4;
    sha256 = "sha256-Eq37qKn5P7rkufJBhME0OG/3BsWlCsFYVnWHP5uVhAQ=";
  };
```
This way, the folder get committed to the nix store. I find this version rather more confusing and error prone than using the `src` argument. What do you think?

doronbehar replied:
> Ok, I gave it a try, but this is actually harder than I thought. The reason is, that the `yarnLock` file is a _singular_ file in the nixstore, completely independent of the project folder. `yarn-berry` needs both files (it actually wants the project folder, but these two files suffice), but we cannot deduce the project folder from a yarn lockfile which is just a store path. We _need_ a directory (or a separate `packageJson` argument)

Oh I think I have a guess what's going here - you were testing changes to the `berry_{3,4}` attributes at `tests/default.nix`? There, a raw file / directory is currently used (e.g `src = ./berry_3`) - Nix puts that file in the store as is, but in the case of regular packages that have a single `src`, that shouldn't happen. Could you push what you did and I'll try to debug that?

gador replied:
sure:
```
diff --git a/pkgs/build-support/node/fetch-yarn-deps/default.nix b/pkgs/build-support/node/fetch-yarn-deps/default.nix
index 90057be587d6..1fe5f1a1906c 100644
--- a/pkgs/build-support/node/fetch-yarn-deps/default.nix
+++ b/pkgs/build-support/node/fetch-yarn-deps/default.nix
@@ -153,7 +153,9 @@ in

               export HOME="$NIX_BUILD_TOP"
               export YARN_ENABLE_TELEMETRY=0
-
+              yarnLock=''${yarnLock:=$PWD/yarn.lock}
+              DIR=$(dirname $yarnLock)
+              cp $DIR/yarn.lock $DIR/package.json $HOME/
               yarn config set enableGlobalCache false
               yarn config set cacheFolder $out
               yarn config set supportedArchitectures --json "$(cat ${./yarn-berry-supported-archs.json})"
diff --git a/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix b/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
index dda298b51a02..d3864be70ea0 100644
--- a/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
+++ b/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
@@ -26,12 +26,12 @@
     sha256 = "sha256-VPnyqN6lePQZGXwR7VhbFnP7/0/LB621RZwT1F+KzVQ=";
   };
   berry_3 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
-    src = ./berry_3;
+    yarnLock = "${./berry_3}/yarn.lock";
     yarnVersion = 3;
     sha256 = "sha256-eiVOpRHuAS7zb6JpCzS3TIHbqQU+wCrMItzZGrE6Fbo=";
   };
   berry_4 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
-    src = ./berry_4;
+    yarnLock = "${./berry_4}/yarn.lock";
     yarnVersion = 4;
     sha256 = "sha256-Eq37qKn5P7rkufJBhME0OG/3BsWlCsFYVnWHP5uVhAQ=";
   };
```

gador replied:
> Oh I think I have a guess what's going here - you were testing changes to the `berry_{3,4}` attributes at `tests/default.nix`? There, a raw file / directory is currently used (e.g `src = ./berry_3`) - Nix puts that file in the store as is, but in the case of regular packages that have a single `src`, that shouldn't happen. Could you push what you did and I'll try to debug that?

You're right, this is probably only needed for this artificial test suite and should work fine in a real work scenario. The docs should probably make a note of that. So everyone knows that pointing `yarnLock` to a store path (file) will fail.

doronbehar replied:
> This way, the folder get committed to the nix store. I find this version rather more confusing and error prone than using the `src` argument. What do you think?

I too think that using `src` for both yarn classic and yarn berry looks much better, but it'd require changing the function signature for current usages of the function - a somewhat tedious task that I don't feel comfortable asking you to perform. Also, I noticed that `nix build -Lf. fixup-yarn-lock.tests.berry_{3,4}` doesn't work with the diff you just sent.

gador replied:
yeah, me too. This works:
```diff
diff --git a/pkgs/build-support/node/fetch-yarn-deps/default.nix b/pkgs/build-support/node/fetch-yarn-deps/default.nix
index 90057be587d6..e6bda1c67038 100644
--- a/pkgs/build-support/node/fetch-yarn-deps/default.nix
+++ b/pkgs/build-support/node/fetch-yarn-deps/default.nix
@@ -153,7 +153,10 @@ in

               export HOME="$NIX_BUILD_TOP"
               export YARN_ENABLE_TELEMETRY=0
-
+              yarnLock=''${yarnLock:=$PWD/yarn.lock}
+              DIR=$(dirname $yarnLock)
+              cp $DIR/yarn.lock $DIR/package.json $HOME/
+              chmod +w $HOME/yarn.lock $HOME/package.json
               yarn config set enableGlobalCache false
               yarn config set cacheFolder $out
               yarn config set supportedArchitectures --json "$(cat ${./yarn-berry-supported-archs.json})"
diff --git a/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix b/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
index dda298b51a02..d3864be70ea0 100644
--- a/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
+++ b/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
@@ -26,12 +26,12 @@
     sha256 = "sha256-VPnyqN6lePQZGXwR7VhbFnP7/0/LB621RZwT1F+KzVQ=";
   };
   berry_3 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
-    src = ./berry_3;
+    yarnLock = "${./berry_3}/yarn.lock";
     yarnVersion = 3;
     sha256 = "sha256-eiVOpRHuAS7zb6JpCzS3TIHbqQU+wCrMItzZGrE6Fbo=";
   };
   berry_4 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
-    src = ./berry_4;
+    yarnLock = "${./berry_4}/yarn.lock";
     yarnVersion = 4;
     sha256 = "sha256-Eq37qKn5P7rkufJBhME0OG/3BsWlCsFYVnWHP5uVhAQ=";
   };
```

doronbehar replied:
OK great I'm glad you cracked it. I'll leave to you the decision of whether to change the `fetchYarnDeps` signature now or later. An intermediate option is to support both signatures at the moment, and create a tracking issue that will recommend all maintainers to manually switch to the new signature. When all Nixpkgs usages have transitioned, we can throw an error if `yarnLock` argument is used, to spread the change in usages outside Nixpkgs.

gador replied:
Ready for another look :+1: 
Right now, the signature is the same as the existing one (so it will work with a supplied `yarnLock` file), but also accepts a `src` argument. 
After this PR gets merged, I can start a tracking issue to recommend users to switch to `src` for all preexisting `fetchYarnDeps` instances

doronbehar replied:
> Right now, the signature is the same as the existing one (so it will work with a supplied `yarnLock` file), but also accepts a `src` argument.
> After this PR gets merged, I can start a tracking issue to recommend users to switch to `src` for all preexisting `fetchYarnDeps` instances

That sounds great, but could you point me please to where the desired `src` signature is treated? I'd like to see if there is a TODO there.

gador replied:
I added a TODO above the function signature

doronbehar replied:
> I added a TODO above the function signature

OK this is good enough in the meantime.

doronbehar commented on doc/languages-frameworks/javascript.section.md:
```diff
@@ -594,6 +594,24 @@ To install the package `yarnInstallHook` uses both `npm` and `yarn` to cleanup p
+
+The hook to be used with `yarn-berry` is called `yarnBerry3ConfigHook` for version 3 lockfiles and `yarnBerry4ConfigHook` for version 4 lockfiles.
+
+There is currently neither an `yarnBuildHook` nor an `yarnInstallHook` for `yarn-berry`.
```
And classic yarn v1 doesn't fit such projects? Perhaps that should be slightly more clear?

gador replied:
I'm not sure what you mean. If upstream uses yarn-v1, we can use the usual hooks. It upstream uses a yarn-berry lockfile (and package.json) we'll have to use either the `yarnBerry3ConfigHook` or the `yarnBerry4ConfigHook`.

gador replied:
got it. Done

doronbehar reviewed (commented):
Only now I read the docs, so I guess 1 more last round of review is needed :pray: .

szlend commented on pkgs/build-support/node/fetch-yarn-deps/yarn-berry-config-hook.sh:
```diff
@@ -0,0 +1,30 @@
+yarnBerryConfigHook(){
+    echo "Executing yarnBerryConfigHook"
+
+    # Use a constant HOME directory
+    mkdir -p /tmp/home
+    export HOME=/tmp/home
```
Sorry if I missed this in another comment, but why is this necessary? Would `$NIX_BUILD_TOP`/`$TMP` not work here?

doronbehar replied:
> Sorry if I missed this in another comment, but why is this necessary? Would `$NIX_BUILD_TOP`/`$TMP` not work here?

Hmm I'm not sure that `NIX_BUILD_TOP` is a good idea, see https://github.com/NixOS/nixpkgs/issues/189691 . And the idea here I think is to use a constant and known temporary directory (not using `mktemp -d`). I don't remember exactly what were the issues I had when I added that in [`a84f3ca` (#318015)](https://github.com/NixOS/nixpkgs/pull/318015/commits/a84f3ca3d824b306ab3080b11495327763d7f2f4) , but I'm open to debate this. Is $TMP an environment variable that is constant for every derivation build?

szlend replied:
Should be constant on Linux, but not necessarily on darwin because darwin doesn't have linux's filesystem namespaces / chroot, so there's no way to ensure the same build dir is free.

I'm not sure about the nix shell issue though. It feels like a bug in nix shell tbh. It should set these env vars appropriately, but I get that might be difficult to fix right now, so happy to keep it /tmp if this is an issue.

I'd still leave a comment linking to the github issue though.

doronbehar replied:
According to https://nix.dev/manual/nix/stable/language/derivations `TMP` should be OK too. Would you be satisfied if instead of `/tmp/home` would be replaced by `$TMP/home`?

gador replied:
Done

doronbehar replied:
> Done

@gador while at it, could you do me a favor and add another commit before the commit titled `fetch-yarn-deps: add yarn-berry support` that will modify the `yarnConfigHook` as well to use `$TMP/home`?

gador replied:
sure

szlend commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -174,6 +210,28 @@ in
     };
   } ./yarn-config-hook.sh;
 
+  yarnBerry3ConfigHook = makeSetupHook {
```
Personally I like how in `pnpm` they put all their hooks under the package itself (e.g. `pnpm_9.configHook`).

This way you can just:
```nix
let
  yarn = yarn-berry_3;
in
{
  yarnOfflineCache = yarn.fetchDeps { ... };
  nativeBuildInputs = [ yarn yarn.configHook ];
}
```

doronbehar replied:
> With your suggestion, it will mean a refactor of all packages using `yarn` to now use `yarn.fetchYarnDeps`.

Small correction: `yarn.fetchDeps` :).

> This can of course be done, but IMHO would be too much for this PR. Lets keep it in mind and maybe do this refactor in a separate PR or tracking issue. This is what @doronbehar suggested for the long-term ecosystem for the `src` argument ([#355053 (comment)](https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1847179151))

Furthermore, this suggestion can ease the refactoring: We could make `yarn.fetchDeps` get the new signature with `src`, and simply make `fetchYarnDeps` throw a message like `fetchYarnDeps is deprecated in favor of yarn.fetchDeps which accepts an src argument instead of yarnLock argument`. This would allow us to not make the effort of making the same function compatible with both signatures.

szlend replied:
I mean we could easily just make these as aliases. But agree that this couls be done in another PR.

gador replied:
Ok, lets target that for another PR

gador committed:
yarn-config-hook: use $TMP instead of /tmp

$TMP should be univsal for all derivations and should
work on darwin as well, if it ever switches to another
build-dir (e.g. /nix/)

see https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1849847398

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

doronbehar commented:
OK most issues are addressed. Let's give @winterqt and @szlend a day or 2 to reply (@winterqt  to [this thread](https://github.com/NixOS/nixpkgs/pull/355053#pullrequestreview-2440805730)).

gador commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -132,22 +138,52 @@ in
-              (cd $out; prefetch-yarn-deps --verbose --builder $yarnLock)
-
-              runHook postBuild
+              DIR=$(dirname $yarnLock)
+              cp $DIR/yarn.lock $DIR/package.json $HOME/
```
Yes I know.
See https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1847112682
But if you read the follow up, this shouldn't be an issue, because we ususally use the `src` of the parent derivation and have the for

the-sun-will-rise-tomorrow replied:
OK. For context I tried using this pull request to package a Yarn project outside nixpkgs, and ran into this problem.

I guess what you're saying is that we should always be using

```nix
yarnLock = src + "/yarn.lock";
```
instead of

```nix
yarnLock = src/yarn.lock;
```

Which is understandable but seems like an easy mistake to make so I think mentioning it in the documentation or having a check would be nice.

the-sun-will-rise-tomorrow replied:
> See [#355053 (comment)](https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1847112682)

Sorry, I'm not sure what's going on with GitHub (issues redesign I guess) but this link isn't working for me, neither by clicking nor opening in a new tab.

gador replied:
Mmh weird, I'm on mobile. Maybe because of that..

You can have at the tests at how I solved that problem with the store path.

Thanks for testing it!

doronbehar replied:
@the-sun-will-rise-tomorrow is correct that the current state in which `yarnLock` is the argument of the function is slightly misleading. Perhaps after all we should change the function signature of every `fetchYarnDeps` variant now? I'm sorry that I don't have time to put an effort in that myself, I'd have loved to help.

gador replied:
I'm with you on the last comment. Lets issue a warning for now, so it catches if only the lock file from the nix store is passed and keep the API the same. In another refactor PR, we can then
1. Move it under the package (e.g. `yarn.fetchDeps` see https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1849863680)
2. Use `src` for `yarn` and `yarn-berry` which will illiminate this current issue

doronbehar commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -132,22 +138,61 @@ in
-              mkdir -p $out
-              (cd $out; prefetch-yarn-deps --verbose --builder $yarnLock)
+              DIR=$(dirname $yarnLock)
+              if [ $DIR = "/nix/store" ]; then
+                echo "Only a yarn.lock file has been passed (e.g. yarnLock = ./yarn.lock;)"
+                echo "Please change this so a directory can be deduced (e.g. yarnLock = finalAttrs.src + "/yarn.lock";)"
+                exit 1
+              fi
```
LGTM.

winterqt reviewed (changes requested):
In general, I am very against adding these kinds of fetchers. (I was too burnt out to comment much on the pnpm situation, and I wish I had the spoons to object more there.)

Upstream makes no guarantees as to the stability of this, and as [we've seen with Rust](https://github.com/NixOS/nixpkgs/pull/217084), this is a ticking time bomb.

As an alternative, we could try to e.g. create our own PnP-based fetcher or something? It takes [a](https://github.com/NixOS/nixpkgs/pull/189539) [lot](https://github.com/NixOS/nixpkgs/pull/200470) [of](https://github.com/NixOS/nixpkgs/pull/206476) [time](https://github.com/NixOS/nixpkgs/pull/206477) to do things properly as various edge cases are discovered, though, and time isn't something I have a lot of at the moment.

*If* this is to be merged in its current state, please address the issues I've commented on in this review. However, I would strongly recommend against doing so.

emilazy commented on pkgs/build-support/node/fetch-yarn-deps/yarn-berry-supported-archs.json:
```diff
@@ -0,0 +1,16 @@
+{
```
This file means that we can never expand our platform support without mass FOD breaks across the tree? That seems very bad. There are already platforms you can boot NixOS on that would be ruled out by this file. At the very least we should try to list every value already used in npm here so that it’s relatively safe to expand in future?

gador replied:
> This file means that we can never expand our platform support without mass FOD breaks across the tree? 

yes

>That seems very bad. There are already platforms you can boot NixOS on that would be ruled out by this file. At the very least we should try to list every value already used in npm here so that it’s relatively safe to expand in future?

We did add all the relevant platforms, cpu's and libc's (see https://yarnpkg.com/configuration/yarnrc#supportedArchitectures) that are available. The only one missing is `win32` and I don't think we need to add this

emilazy replied:
If it’s everything Yarn supports then I guess it’s not so bad – assuming we will be able to add additional platforms if Yarn does support more in future without invalidating every FOD?

I do think we should include Windows. You can use Nixpkgs to cross‐build a lot of stuff for Windows today, and people are working on native bring‐up. What’s the harm?

gador replied:
ok, lets add `win32`.
If yarn adds support for further platforms, we can add those, but we will need to re-validate all of the hashes involved. Hopefully this will be very rarely the case..

emilazy replied:
I thought the idea is that packages can’t include any platforms other than the ones Yarn supports, so no existing FOD should change as long as we keep up with Yarn? Or can packages on npm reference other platforms and Yarn just can’t support fetching them?

gador replied:
done. 
On a sidenode: I did neither have to change the test hashes, nor the hash for the (seriously complex and long) yarn pgadmin offlineCache. (and yes, I obviously invalidated them, re-downloaded the packages and compared the hashes)
So maybe this won't be such a big deal after all

gador replied:
> I thought the idea is that packages can’t include any platforms other than the ones Yarn supports, so no existing FOD should change as long as we keep up with Yarn? 

Yes, I think your right. My guess is that adding another platform or compiler will a huge undertaking anyway, so maybe they will publish yet another big version upgrade, which would make it easier for us (e.g. `yarn-berry_5`)

emilazy replied:
Unfortunately, I think I was correct here:

> Or can packages on npm reference other platforms and Yarn just can’t support fetching them?

The npm docs imply that the fields are quite freeform; e.g. the [`cpu` example](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#cpu) uses `mips` as an example. That means that this is just a limitation of Yarn’s configuration format rather than a constraint on the actual packages in existence, and there are packages that could change in the future if we add new platforms like RISC‐V.

The correct solution here would be to fetch everything independent of platform – theoretically Yarn could add `all` values here to opt in to fetching everything, but writing our own deterministic fetcher would make it easy and also obviate the other reproducibility concerns. If neither of those is an option, perhaps we can find a way to check that nothing in the dependency tree references any platform we haven’t included in this file, and otherwise bail out? If we can’t do even that, then it’d be good if someone could write a script to scan the npm repository and collect every value that is used in practice so that we can include them all from the start and reduce the scope of potential changes.

Because we’re adding fundamental packaging infrastructure here that is part of our public interface and will surely be relied upon by out‐of‐tree packages, and changing FOD hashes are very difficult to spot and debug, I don’t think we’ll have the option of *just breaking it* in the future, unless we can show that the backwards incompatibility would be entirely theoretical in the existing public corpus.

gador replied:
> Unfortunately, I think I was correct here:
> 
> > Or can packages on npm reference other platforms and Yarn just can’t support fetching them?
> 
> The npm docs imply that the fields are quite freeform; e.g. the [`cpu` example](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#cpu) uses `mips` as an example. That means that this is just a limitation of Yarn’s configuration format rather than a constraint on the actual packages in existence, and there are packages that could change in the future if we add new platforms like RISC‐V.
> 

Actually, it is not a limitation of the configuration, as it can be adapted to include everything node supports. https://yarnpkg.com/configuration/yarnrc#supportedArchitectures says:

>List of CPU architectures to cover.
See https://nodejs.org/docs/latest/api/process.html#processarch for the architectures supported by Node.js

which states:
>The operating system CPU architecture for which the Node.js binary was compiled. Possible values are: 'arm', 'arm64', 'ia32', 'loong64', 'mips', 'mipsel', 'ppc', 'ppc64', 'riscv64', 's390', 's390x', and 'x64'.

So we can include all of those for our `supported-archs`. Adding this, btw, didn't change the cache for pgadmin, either.
I looked into how `yarn-berry` does its comparisons (since these values are for the local `.yarnrc.toml` config file) and it compares it to upstreams `package.json` file. This file has the same fields (https://yarnpkg.com/configuration/manifest#cpu) and, more importantly, _only_ the `libc` and `os` fields we already have.

> The correct solution here would be to fetch everything independent of platform – theoretically Yarn could add `all` values here to opt in to fetching everything, but writing our own deterministic fetcher would make it easy and also obviate the other reproducibility concerns. 

Yes, writing our own fetcher could be a solution. This will be quite a task to parse the `yarn.lock` file and generate the cache. I really feel uncomftable reinventing the wheel, though. We do have a package manager, which already parses this file and also controlls the specs about our architectures. I don't want to write yet another solution of `yarn2nix`

>If neither of those is an option, perhaps we can find a way to check that nothing in the dependency tree references any platform we haven’t included in this file, and otherwise bail out? 

There seems to be an option (`yarn info --all --manifest --json`) which prints all the manifests, but _not one_ had a `cpu`, `libc` or `os` field. There is another option (`yarn npm info _packageName_ --json`) which prints the upstream manifest and from the quick peek I took, didn't include any of those, either.

>If we can’t do even that, then it’d be good if someone could write a script to scan the npm repository and collect every value that is used in practice so that we can include them all from the start and reduce the scope of potential changes.
> Because we’re adding fundamental packaging infrastructure here that is part of our public interface and will surely be relied upon by out‐of‐tree packages, and changing FOD hashes are very difficult to spot and debug, I don’t think we’ll have the option of _just breaking it_ in the future, unless we can show that the backwards incompatibility would be entirely theoretical in the existing public corpus.

Totally agree. This is why we added versions here. Not only because upstream has two supported versions (3 and 4), but also because their `yarn.lock` files differ. 
If there is a situation where either the file format changes, or the supported architectures, I would suggest adding another version (e.g. `yarn-berry_5` or `yarn-berry_39`). This will keep all the hashes intakt, be reproducible and still be able to adapt to future changes.

szlend replied:
> The npm docs imply that the fields are quite freeform; e.g. the [`cpu` example](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#cpu) uses `mips` as an example. That means that this is just a limitation of Yarn’s configuration format rather than a constraint on the actual packages in existence, and there are packages that could change in the future if we add new platforms like RISC‐V.

The documentation says these values are based on Node `process.{arch,cpu}` which is a known [documented](https://nodejs.org/api/process.html#processarch) set of values. I wouldn't exactly call it freeform.

I don't really have a better idea on how to handle adding new platform other than invalidating the FOD (by changing `pname`) and fixing up all the builds or introducing a new builder version.

I wonder how `pnpm` managed to avoid this issue. I don't see it handling `supportedArchitectures` at all.

gador commented on pkgs/build-support/node/fetch-yarn-deps/yarn-config-hook.sh:
```diff
@@ -2,8 +2,8 @@ yarnConfigHook(){
     # Use a constant HOME directory
-    mkdir -p /tmp/home
-    export HOME=/tmp/home
+    mkdir -p $TMP/home
+    export HOME=$TMP/home
```
See https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1850230901 and https://nix.dev/manual/nix/2.18/language/derivations
$TMP = $TMPDIR = $TEMPDIR = $TEMP

I'll add the quotes 👍

gador commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -132,22 +138,61 @@ in
-              mkdir -p $out
-              (cd $out; prefetch-yarn-deps --verbose --builder $yarnLock)
+              DIR=$(dirname $yarnLock)
+              if [ $DIR = "/nix/store" ]; then
```
good point

gador commented on pkgs/development/tools/yarn-berry/default.nix:
```diff
@@ -5,17 +5,25 @@
 }:
 
+let
+  version_4 = "4.5.1";
+  version_3 = "3.8.6";
```
done

gador commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -132,22 +138,61 @@ in
+              (
+                {
+                  "1" = prefetch-yarn-deps;
+                  "3" = yarn-berry_3;
+                  "4" = yarn-berry_4;
```
I added it at the package level, because this is where we would do our update

doronbehar commented on pkgs/build-support/node/fetch-yarn-deps/default.nix:
```diff
@@ -132,22 +138,61 @@ in
+
+                mkdir -p $out
+              ''
+              + lib.optionalString (yarnVersion > 1) ''
```
> Can we add a diagnostic that attempts to detect whether the `yarnVersion` is correct?

If I understand correctly the current state of things, such a check wouldn't make a real difference, as setting `yarnVersion` should make the `fetchYarnDeps` function use already a certain `yarn` derivation in the `nativeBuildInputs`.. So is it meant to simply check [this logic](https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1836939084)?

gador replied:
No.
It checks the content of the yarn file and checks whether it the the correct version for the correct tool (__metadata is only found in yarn-berry lock files)

doronbehar replied:
> No. It checks the content of the yarn file and checks whether it the the correct version for the correct tool (__metadata is only found in yarn-berry lock files)

Oh I see, makes sense! Thanks for explaining :).

gador commented:
@emilazy I know this is a controversial PR, but I'd love to here your thoughts on https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1855133116 
thanks :-)

gador commented:
@emilazy I see that there is no activity from you since end of November. I'd still love to get your feedback to my reply above. In the meantime: @SuperSandro2000 could you also have a look at this? I'd like to get this merged and I believe I addressed most of the issues concerning stability and reliability in [my comment](https://github.com/NixOS/nixpkgs/pull/355053#discussion_r1855133116) above

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-implement-version-2-yarn-lock-package-management/59792/1

yuyuyureka reviewed (changes requested):
Sadly, I have to agree with winterqt.  
This kind of fetcher is not sustainable and just defers the cost of an actual pure fetcher to the future.  
The reason is ultimately that using FODs like this is always a hack, and the only valid use of them is to reduce evaluation cost and avoid Import-From-Derivation for a problem that we could theoretically also solve by generating many small derivations which would match the hashes specified in the lockfile.  
If we don't really understand what yarn is doing under the hood to arrive at this hash, it will explode in our faces sooner or later.

flokli commented:
https://github.com/NixOS/nixpkgs/pull/399404 has landed, this can be closed.
